packages/ckb-sdk-core/src/signWitnessGroup.ts

```diff
@@ -39,10 +57,18 @@ const signWitnessGroup = (
   if (typeof sk === 'string') {
     const keyPair = new ECPair(sk)
     emptyWitness.lock = keyPair.signRecoverable(message)
+    return [multisigConfig ? emptyWitness : serializeWitnessArgs(emptyWitness), ...witnessGroup.slice(1)]
   } else {
-    emptyWitness.lock = sk(message)
+    const skResult = sk(message, [emptyWitness, ...witnessGroup.slice(1)])
+    if (typeof skResult === 'string') {
+      emptyWitness.lock = skResult
+      return [multisigConfig ? emptyWitness : serializeWitnessArgs(emptyWitness), ...witnessGroup.slice(1)]
+    }
+    return skResult.then(res => {
```

Comment:
If the result of sk function is string, it's synchronous.

Reply:
If the result of sk function is string, it's synchronous.

Reply:
the promise `skResult` will infect functions that invoke `signWitnessGroup`

Reply:
In which case `skResult` should be a promise

Reply:
It will be a promise with hard wallet.
`
function signWitnessGroup(
  sk: SignatureProvider,
  transactionHash: TransactionHash,
  witnessGroup: StructuredWitness[],
  multisigConfig?: MultisigConfig
): StructuredWitness[]
`
`
function signWitnessGroup(
  sk: (message: string | Uint8Array, witness: StructuredWitness[]) => Promise<string>,
  transactionHash: TransactionHash,
  witnessGroup: StructuredWitness[],
  multisigConfig?: MultisigConfig
): Promise<StructuredWitness[]>
`

I defined two declarations. So whether to return promise is controlled by the caller

Reply:
Gotcha